src/process/ProcessManagerImpl.cpp

```diff
@@ -564,6 +567,11 @@ ProcessManagerImpl::handleSignalWait()
             const int status = std::get<1>(pidStatus);
             handleProcessTermination(pid, status);
         }
+        if (mSyncFilesystemOnProcessExit)
```

Comment:
Should this be done before `handleProcessTermination` ?

Also, this seems too aggressive and unsafe:

* I *think* this gets executed off the main thread, and will stall it for arbitrary long
* even if we can push this to the worker thread, the main thread's simple writes (like logging) will get stalled behind all pending I/O.

I don't know if it's possible, but as we know the output filename, can't we invoke the SubCommand followed by something like (executed in a thread/or some clever async fsync):

```
fd = open(fn);
fsync(fd);
close(fd);
```

before claiming that the process completed and calling the `handleProcessTermination`

Obviously the potential issue here is that this doesn't work as `fd` doesn't get any new writes but maybe there is a way to trick the kernel in doing it, like append 1 bytes + truncate 1 byte or something like that.

Reply:
Ah found it, there is `syncfs` that does what we want. It's what happens when invoking from a shell the command
```
sync -f somefile
```

Reply:
It does indeed get executed on the main thread -- that's the whole point, to block the main thread from doing more allocation / launching more processes / causing more VM pressure until existing dirty pages are written.

I'm not entirely sure what you're concerned about here: flushing too many filesystems besides the one we're intending to? Or flushing files on that filesystem that aren't necessary to flush?

fsync on a file flushes changes to that file -- we certainly could _just_ fsync the file that was transferred by (say) each get() command, that might be sufficient to relieve the pressure we're encountering, might not. In theory it should be but it's hard to tell.

I _think_ (trying to read between the lines of the man page) that opening a recently-written file -- even one written by a different process -- then calling `fsync` and closing it will in fact flush dirty pages associated with that file to the disk. I mean, I can try it and see anyways. If you prefer to do this narrower single-file version, I will try that.

`syncfs` is the same as `sync` but on a single filesystem. So it's a bit in-between `sync` (all filesystems) and `fsync` (one file). I'm not sure it's worth bothering to split that hair, since we're probably running on a single-filesystem node anyways (or only have one filesystem with much dirty data). Also `syncfs` isn't portable to macos or BSD.

Reply:
no, what I am worried about is that the current behavior with `RunCommand` is that all work happens in a different process and the "join" with the main thread is virtually free as it happens when the child process finishes.

With that flag enabled, the sub process creates a bunch of debt that is paid back by the main thread. This can block for arbitrary long and will cause all sorts of havok in our single threaded app (like kill connections etc).

Then there is using system wide sync:
* some commands that don't even write to disk will potentially hang for a long time (for example, publish), as it may block on unrelated IO like a bucket merge or random log aggregator like the one that we have running on our boxes.
* this may negatively impact other processes running in the same container (or maybe it's even broader than this, like the whole host VM). I don't know how bad it can get for postgres or Horizon (captive core for sure is running in the same container).